### PR TITLE
fix(pp): fix no-packed dataset load micro batch error

### PR DESCRIPTION
**Pull request by SolenoidWGT**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/core/engine.py [hunk 1]
```diff
@@ -185,6 +185,11 @@ def load_batch(self, data_iter, to_gpu=True):
 
         if to_gpu:
             batch_data = move_to_device(batch_data)
+
+        # For packed-dataset, batch_data is (micro_num, micro_num*micro_bsz),
```
Comment (yingtongxiong) [inline internlm/core/engine.py:189]: “For packed-dataset, batch_data is (micro_num, micro_bsz*seq_len)” 这里应该是这样？

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 2]
```diff
@@ -186,17 +190,25 @@ def _call_engine(engine, data):  # pylint: disable=W0237
             raise TypeError(f"Expected data to be of type torch.Tensor, list, tuple, or dict, but got {type(data)}")
 
     def load_batch(self, engine, data_iter):
-        # Pipeline schedule just puts data in memory
+        # Pipeline schedule just puts data in memory,
         batch_data, actual_batch_size = engine.load_batch(data_iter, to_gpu=False)
 
-        self.num_microbatches = actual_batch_size  # Rampup or variable bsz size.
+        if check_data_is_packed(batch_data):
```
Comment (yingtongxiong) [inline internlm/core/scheduler/pipeline_scheduler.py:196]: 这个函数的名称或许可以改一下？或者加个注释？因为我发现，当我们把use_flash_attn设置为False的时候，在训练的时候，batch_data里面依然有"indexes"这个key，然后在evaluation的时候才没有“indexes”这个key？

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 3]
```diff
@@ -186,17 +190,25 @@ def _call_engine(engine, data):  # pylint: disable=W0237
...
+        else:
+            micro_num = actual_batch_size // gpc.config.data["micro_bsz"]
+
         self.microbatch_offset = 0
         self.batch_size = actual_batch_size
         self.batch_data, self.batch_label = batch_data
+        self.bsz_stride = self.batch_size // micro_num
+        # 'num_microbatches' is no longer an initialization parameter,
+        # but is determined on the fly by the Scheduler.
+        self.num_microbatches = micro_num  # Rampup or variable bsz size.
```
Comment (yingtongxiong) [inline internlm/core/scheduler/pipeline_scheduler.py:207]: 想再确认一下，evaluation应该是不受rampup影响？

Comment (SolenoidWGT) [reply internlm/core/scheduler/pipeline_scheduler.py:207]: 不受影响的，其实这里的改动就是不再把num_microbatches作为一个初始化参数，而是on the fly的从batch的维度判断，这样就可以只修改sampler来支持动态的bsz

File: internlm/core/scheduler/pipeline_scheduler.py (hunk 2 above)
Comment (SolenoidWGT) [reply internlm/core/scheduler/pipeline_scheduler.py:196]: 已补充注释

File: internlm/core/engine.py (hunk 1 above)
Comment (SolenoidWGT) [reply internlm/core/engine.py:189]: 已修改

**Review by yingtongxiong — APPROVED** [review]

**Pull request merged**